mypy/build.py

```diff
@@ -962,6 +962,77 @@ def verify_module(fscache: FileSystemCache, id: str, path: str) -> bool:
     return True
 
 
+def write_protocol_deps_cache(proto_deps: Dict[str, Set[str]],
+                              manager: BuildManager, graph: Graph) -> None:
+    """Write cache files for protocol dependencies.
+
+    Serialize protocol dependencies map for fine grained mode. Also take the snapshot
+    of current sources to later check consistency between protocol cache and individual
+    cache files.
+    """
+    proto_meta, proto_cache = get_protocol_deps_cache_name(manager)
+    meta_snapshot = {}  # type: Dict[str, str]
+    error = False
+    for id in graph:
+        meta_snapshot[id] = graph[id].source_hash
+    if not atomic_write(proto_meta, json.dumps(meta_snapshot), '\n'):
+        manager.log("Error writing protocol meta JSON file {}".format(proto_cache))
+        error = True
+    listed_proto_deps = {k: list(v) for (k, v) in proto_deps.items()}
+    if not atomic_write(proto_cache, json.dumps(listed_proto_deps), '\n'):
+        manager.log("Error writing protocol deps JSON file {}".format(proto_cache))
+        error = True
+    if error:
+        manager.errors.set_file(_get_prefix(manager), None)
+        manager.errors.report(0, 0, "Error writing protocol dependencies cache.",
+                              blocker=True)
+
+
+def read_protocol_cache(manager: BuildManager,
+                        graph: Graph) -> Optional[Dict[str, Set[str]]]:
+    """Read and validate protocol dependencies cache."""
+    proto_meta, proto_cache = get_protocol_deps_cache_name(manager)
+    try:
+        data = manager.fscache.read(proto_meta).decode()
+        manager.trace('Proto meta {}'.format(data.rstrip()))
+        meta_snapshot = json.loads(data)
+    except IOError:
+        manager.log('Could not load protocol meta: could not find {}'.format(proto_meta))
```

Comment:
`IOError` can be raised due to various reasons even if the file exists, so the message "could not find ..." may not be accurate. Maybe just log something like "Could not load protocol metadata: <path>".